planner/core/rule_column_pruning.go

```diff
@@ -658,3 +651,57 @@ func appendItemPruneTraceStep(p LogicalPlan, itemType string, prunedObjects []fm
 	}
 	opt.appendStepToCurrent(p.ID(), p.TP(), reason, action)
 }
+
+// pick a not null and shortest column from table
+func preferNotNullColumnFromTable(dataSource *DataSource) (*expression.Column, *model.ColumnInfo) {
+	var resultColumnInfo *model.ColumnInfo
+	var resultColumn *expression.Column
+	for _, columnInfo := range dataSource.tableInfo.Columns {
+		if columnInfo.Hidden || columnInfo.FieldType.IsVarLengthType() {
+			continue
+		}
+		if mysql.HasNotNullFlag(columnInfo.GetFlag()) {
+			if resultColumnInfo == nil || columnInfo.GetFlen() < resultColumnInfo.GetFlen() {
+				resultColumnInfo = columnInfo
+				resultColumn = &expression.Column{
+					UniqueID: dataSource.ctx.GetSessionVars().AllocPlanColumnID(),
+					ID:       resultColumnInfo.ID,
+					RetType:  resultColumnInfo.FieldType.Clone(),
+					OrigName: fmt.Sprintf("%s.%s.%s", dataSource.DBName.L, dataSource.tableInfo.Name.L, resultColumnInfo.Name),
+					IsHidden: resultColumnInfo.Hidden,
+				}
+			}
+		}
+	}
+	if resultColumnInfo == nil {
+		// todo
+		if dataSource.handleCols != nil {
+			resultColumn = dataSource.handleCols.GetCol(0)
+			resultColumnInfo = resultColumn.ToInfo()
+		} else {
+			resultColumn = dataSource.newExtraHandleSchemaCol()
+			resultColumnInfo = model.NewExtraHandleColInfo()
+		}
+	}
+	return resultColumn, resultColumnInfo
+}
+
+func preferKeyColumnFromTable(dataSource *DataSource, originColumns []*expression.Column,
+	originSchemaColumns []*model.ColumnInfo) (*expression.Column, *model.ColumnInfo) {
+	var resultColumnInfo *model.ColumnInfo
+	var resultColumn *expression.Column
+	if dataSource.table.Type().IsClusterTable() && len(originColumns) > 0 {
+		// use the first line.
```

Comment:
first column?

Reply:
This is a comment on his previous logic, I just moved the position. I understand what he meant, if the table is a clustered table, it will use the first column.

Reply:
ok, then just update the comment.